File: backend/src/services/ai/aiService.js
export function getAiConfig() {
  const apiKey =
    process.env.AI_API_KEY ||
    process.env.GEMINI_API_KEY ||
    process.env.GOOGLE_API_KEY ||
    process.env.OPENAI_API_KEY ||
    process.env.DASHSCOPE_API_KEY ||
    '';
  const baseUrl =
    process.env.AI_BASE_URL ||
    'https://generativelanguage.googleapis.com/v1beta/openai';
  const model =
    process.env.AI_MODEL ||
    'gemini-1.5-flash';

  return { apiKey, baseUrl, model };
}

export function isAiAvailable() {
  const { apiKey } = getAiConfig();
  return Boolean(apiKey && apiKey.trim().length > 0);
}

export async function callAiChat(messages, { temperature = 0.3 } = {}) {
  const { apiKey, baseUrl, model } = getAiConfig();

  if (!isAiAvailable()) {
    console.warn('[HumSaathi AI] No AI API Key found in environment variables (checked AI_API_KEY, GEMINI_API_KEY, GOOGLE_API_KEY, OPENAI_API_KEY). Using fallback.');
    return null;
  }

  try {
    const response = await fetch(`${baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({
        model,
        temperature,
        messages,
        response_format: { type: 'json_object' },
      }),
    });

    if (!response.ok) {
      const errText = await response.text();
      console.error(`[HumSaathi AI Error] HTTP ${response.status} (${response.statusText}):`, errText);
      return null;
    }

    const data = await response.json();
    const content = data.choices?.[0]?.message?.content;
    if (!content) {
      console.warn('[HumSaathi AI] Empty content in response:', data);
      return null;
    }
    return JSON.parse(content);
  } catch (error) {
    console.error('[HumSaathi AI] AI call failed:', error.message);
    return null;
  }
}

export const AI_API_KEY = process.env.AI_API_KEY || process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY || process.env.OPENAI_API_KEY || process.env.DASHSCOPE_API_KEY || '';

File: backend/app/services/ai/ai_service.py
import json
import logging
import re
import httpx
from typing import List, Dict, Any, Optional
from app.config import settings

logger = logging.getLogger("humsaathi-ai")

def is_ai_available() -> bool:
    return bool(settings.AI_API_KEY and settings.AI_API_KEY.strip())

def clean_json_text(raw_text: str) -> str:
    """Strips markdown code fences and whitespace from raw JSON response."""
    text = raw_text.strip()
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*", "", text)
        text = re.sub(r"\s*```$", "", text)
    return text.strip()

async def call_ai_chat(messages: List[Dict[str, str]], temperature: float = 0.5) -> Optional[Dict[str, Any]]:
    """Calls AI for structured JSON responses (evaluation, scoring, scenario feedback)."""
    if not is_ai_available():
        logger.info("[HumSaathi AI] No API key provided. Using rule-based fallback.")
        return None

    url = f"{settings.AI_BASE_URL.rstrip('/')}/chat/completions"
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {settings.AI_API_KEY}",
    }
    payload = {
        "model": settings.AI_MODEL,
        "temperature": temperature,
        "messages": messages,
        "response_format": {"type": "json_object"},
    }

    try:
        async with httpx.AsyncClient(timeout=25.0) as client:
            response = await client.post(url, headers=headers, json=payload)
            if response.status_code != 200:
                logger.warning(f"[HumSaathi AI Error] HTTP {response.status_code}: {response.text}")
                return None

            data = response.json()
            choices = data.get("choices", [])
            if not choices:
                return None
            content = choices[0].get("message", {}).get("content")
            if not content:
                return None

            cleaned = clean_json_text(content)
            try:
                parsed = json.loads(cleaned)
                if isinstance(parsed, dict):
                    return parsed
            except Exception:
                if cleaned:
                    return {"response": cleaned}

            return None
    except Exception as e:
        logger.warning(f"[HumSaathi AI] AI request failed: {e}. Falling back to rules.")
        return None

async def call_ai_text(messages: List[Dict[str, str]], temperature: float = 0.7) -> Optional[str]:
    """Calls AI for natural conversational responses, explanations, code generation, and general Q&A."""
    if not is_ai_available():
        logger.info("[HumSaathi AI] No API key provided. Using contextual knowledge engine.")
        return None

    url = f"{settings.AI_BASE_URL.rstrip('/')}/chat/completions"
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {settings.AI_API_KEY}",
    }
    payload = {
        "model": settings.AI_MODEL,
        "temperature": temperature,
        "messages": messages,
    }

    try:
        async with httpx.AsyncClient(timeout=25.0) as client:
            response = await client.post(url, headers=headers, json=payload)
            if response.status_code != 200:
                logger.warning(f"[HumSaathi AI Text Error] HTTP {response.status_code}: {response.text}")
                return None

            data = response.json()
            choices = data.get("choices", [])
            if not choices:
                return None
            content = choices[0].get("message", {}).get("content")
            if not content:
                return None

            return content.strip()
    except Exception as e:
        logger.warning(f"[HumSaathi AI] AI text request failed: {e}. Falling back to contextual knowledge engine.")
        return None


